crates/task-impls/src/quorum_vote/handlers.rs

```
@@ -266,72 +212,97 @@ async fn start_drb_task<TYPES: NodeType, I: NodeImplementation<TYPES>, V: Versio
 ///
 /// Special cases:
 /// * Epoch 0: No DRB computation since we'll transition to epoch 1 immediately.
 /// * Epoch 1 and 2: No computed DRB result since when we first start the computation in epoch 1,
-///   the result is for epoch 3.
+///     the result is for epoch 3.
 ///
-/// We don't need to handle the special cases explicitly here, because the first proposal with
-/// which we'll start the DRB computation is for epoch 3.
+/// We don't need to handle the special cases explicitly here, because the first leaf with which
+/// we'll start the DRB computation is for epoch 3.
 async fn store_drb_seed_and_result<TYPES: NodeType, I: NodeImplementation<TYPES>, V: Versions>(
-    proposal: &QuorumProposal2<TYPES>,
     task_state: &mut QuorumVoteTaskState<TYPES, I, V>,
-    leaf_views: &[LeafInfo<TYPES>],
+    decided_leaf: &Leaf2<TYPES>,
 ) -> Result<()> {
-    // This is never none if we've reached a new decide, so this is safe to unwrap.
-    let decided_block_number = leaf_views
-        .last()
-        .unwrap()
-        .leaf
-        .block_header()
-        .block_number();
+    let decided_block_number = decided_leaf.block_header().block_number();
 
-    // Skip if this is not the expected block.
-    if task_state.epoch_height != 0 && is_epoch_root(decided_block_number, task_state.epoch_height)
-    {
-        // Cancel old DRB computation tasks.
+    if task_state.epoch_height != 0 {
         let current_epoch_number = TYPES::Epoch::new(epoch_from_block_number(
             decided_block_number,
             task_state.epoch_height,
         ));
 
-        let mut consensus_writer = task_state.consensus.write().await;
-        consensus_writer
-            .drb_seeds_and_results
-            .garbage_collect(current_epoch_number);
-        drop(consensus_writer);
+        // Skip storing the seed and computed result if this is not the epoch root.
+        if is_epoch_root(decided_block_number, task_state.epoch_height) {
```

can this also be an `ensure!()` to avoid the nested `if`s?

---

Similar to the reply above, I don't think we should return an error if this isn't an epoch root, because it may be the last block instead which will be handled later in this function. Does this make sense?